// ngn/shared/tasks/task.js
/**
  * v0.3.31 generated on: Thu May 18 2017 01:24:44 GMT+0000 (UTC)
  * Copyright (c) 2014-2017, Ecor Ventures LLC. All Rights Reserved. See LICENSE (BSD3).
  */
'use strict'

NGN.Task = function () {
  class QueueItem extends NGN.EventEmitter {
    constructor (config) {
      config = config || {}

      super(config)

      Object.defineProperties(this, {
        name: NGN.const(NGN.coalesce(config.name, 'Unknown')),

        callback: NGN.privateconst(config.callback),

        number: NGN.const(parseInt(config.number, 10)),

        timer: NGN.private(null),
        _status: NGN.private(null),
        bus: NGN.private(config.buz),
        _skip: NGN.private(false)
      })

      this.on('timeout', () => {
        this._status = 'timedout'
      })

      this.on('skip', () => {
        this._status = 'skipped'
      })
    }

    get status () {
      return this._status
    }

    get skipped () {
      return this._skip
    }

    run (mode) {
      if (this.skipped) {
        this.emit('skip', this)

        if (mode && mode === 'dev') {
          console.warn('SKIPPED ' + this.name)
        }

        return
      }

      this.emit('start', this)

      if (mode && mode === 'dev') {
        console.info('Executing ' + this.name + ':')
      }

      this._status = 'running'

      const me = this
      const scope = {
        name: this.name,
        number: this.number,
        timeout: function (milliseconds) {
          me.timer = setTimeout(function () {
            me.emit('timeout', me)
          }, milliseconds)
        }
      }

      this.callback.apply(scope, [function () {
        me._status = 'complete'
        me.emit('complete', me)
      }])

      if (this.callback.length === 0) {
        me._status = 'complete'
        me.emit('complete', me)
      }
    }

    skip () {
      if (this._status === 'running') {
        console.warn(`Cannot skip step: ${this.name} is currently running.`)
      } else if (this._status === 'timedout') {
        console.warn(`Cannot skip step: ${this.name} timed out.`)
      } else if (this._status === 'complete') {
        console.warn(`Cannot skip step: ${this.name} already completed.`)
      }

      this._skip = true
    }
  }

  return QueueItem
}

NGN.Task = NGN.Task()

if (NGN.nodelike) {
  module.exports = NGN.Task
}
