// ngn/shared/exception.js
/**
  * v0.3.31 generated on: Thu May 18 2017 01:24:44 GMT+0000 (UTC)
  * Copyright (c) 2014-2017, Ecor Ventures LLC. All Rights Reserved. See LICENSE (BSD3).
  */
'use strict'

class CustomException extends Error { 
  constructor (config) {
    super()

    config = config || {}
    config = typeof config === 'string' ? { message: config } : config
    config.custom = config.custom || {}

    let me = this

    this.name = config.name || 'NgnError'
    this.type = config.type || 'TypeError'
    this.severity = config.severity || 'minor'
    this.message = config.message || 'Unknown Error'
    this.category = config.category || 'operational' 

    this.name = this.name.replace(/[^a-zA-Z0-9_]/gi, '')

    for (var attr in config.custom) {
      if (config.custom.hasOwnProperty(attr)) {
        this[attr] = config.custom[attr]
      }
    }
    this.hasOwnProperty('custom') && delete this.custom

    if (NGN.nodelike || Error.prepareStackTrace) {
      Error.prepareStackTrace = function (_, stack) { return stack }

      let _err = new Error()
      Error.captureStackTrace(_err, this)

      this.rawstack = _err.stack

      Error.prepareStackTrace = function (err, stack) { 
        me.cause && console.warn(me.cause)
        me.help && console.info(me.help)

        return me.name + ': ' + me.message + '\n' + stack.filter(function (frame) {
          return frame.getFileName() !== __filename && frame.getFileName()
        }).map(function (el) {
          return '    at ' + el
        }).join('\n')
      }

      Error.captureStackTrace(this)
    }
  }

  get trace () {
    return this.rawstack.filter(function (frame) {
      return frame.getFileName() !== __filename && frame.getFileName()
    }).map(function (frame) {
      return {
        filename: frame.getFileName(),
        line: frame.getLineNumber(),
        column: frame.getColumnNumber(),
        functionname: frame.getFunctionName(),
        native: frame.isNative(),
        eval: frame.isEval(),
        type: frame.getTypeName()
      }
    })
  }
}

if (NGN.nodelike) {
  module.exports = CustomException
}
